compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.cpp

```diff
@@ -212,40 +205,105 @@ Operation *CoalescedGatherDMAOp::getIteratingParent() {
+  }
+  if (hasMemRef && !isa<MemRefType>(sourceType)) {
+    return emitOpError("source must be memref when init is memref");
+  }
+
+  auto indices = getIndices();
+
+  if (indices.size() > initShape.size()) {
+    return emitOpError("number of indices (")
+           << indices.size() << ") cannot exceed destination rank ("
+           << initShape.size() << ")";
+  }
+
+  if (indices.size() > sourceShape.size()) {
+    return emitOpError("number of indices (")
+           << indices.size() << ") cannot exceed source rank ("
+           << sourceShape.size() << ")";
+  }
+
+  // Make sure indices have no dynamic shapes.
+  for (auto [i, indexVal] : llvm::enumerate(indices)) {
+    auto indexType = cast<ShapedType>(indexVal.getType());
+    for (auto dim : indexType.getShape()) {
+      if (ShapedType::isDynamic(dim)) {
+        return emitOpError("expected index ") << i << " to have static shape";
+      }
+    }
+  }
+
+  // For gather operations with indices, all index vectors should have the same
+  // length equal to the batch size (first dimension of destination).
+  if (!indices.empty()) {
+    // Verify all index vectors are 1D and have the same length
+    auto firstIndexShape = cast<ShapedType>(indices[0].getType()).getShape();
+    if (firstIndexShape.size() != 1) {
+      return emitOpError("expected index 0 to be a 1-D tensor or vector");
+    }
+    int64_t batchSize = firstIndexShape.front();
+
+    for (auto [i, indexVal] : llvm::enumerate(indices)) {
+      auto indexShape = cast<ShapedType>(indexVal.getType()).getShape();
+      if (indexShape.size() != 1) {
+        return emitOpError("expected index ")
+               << i << " to be a 1-D tensor or vector";
+      }
+      if (indexShape.front() != batchSize) {
+        return emitOpError(
+                   "expected all index vectors to have the same length; ")
+               << "index " << i << " has length " << indexShape.front()
+               << " but expected " << batchSize;
+      }
+    }
+
+    // The batch size should match the first dimension of the destination
+    if (!initShape.empty() && batchSize != initShape[0]) {
+      return emitOpError("expected batch size (length of index vectors: ")
+             << batchSize << ") to match first destination dimension ("
+             << initShape[0] << ")";
     }
-  } else if (hasMemRef) {
-    if (!isa<MemRefType>(getIndices().getType()) ||
-        !isa<MemRefType>(getSource().getType())) {
-      return emitOpError("all operands must be memrefs when init is a memref");
+  }
+
+  // Verify the contiguous (non-indexed) dimensions match between source and
+  // dest
+  for (auto [dim, size] : llvm::enumerate(initShape)) {
+    if (dim >= sourceShape.size()) {
+      return emitOpError("expected source to have at least ")
+             << (dim + 1) << " dimensions when destination has rank "
+             << initShape.size();
+    }
+
+    // Skip indexed dimensions - they're validated above
```

Comment:
nit: Punctuate, here and above.
```suggestion
    // Skip indexed dimensions - they're validated above.
```